Refactoring to allow for different signature schemes

LindaOrtega opened the pull request:
Refactoring to prepare for #118 

**MOTIVATION:** 
To allow users to sign transactions with different signature schemes (i.e. ETH/BTC's ECDSA), Pact's `scheme` needs to be expanded to include sections for the **signature algorithm** being used **[1]**, the **hashing algorithm** used to hash the transaction _before_ signing or validating it **[2]**, and the **address (ownership) format** the specific blockchain uses **[3]**. Moreover, incorporating the hash algorithm along with the signature algorithm in `scheme` will allow for flexibility in the future, should we wish to switch away from Blake2b_512 as Pact's default hash for Crypto. 

**[1]** For example, `ECDSA` vs `ED25519`
**[2]** Haskell's ECDSA [signing](http://hackage.haskell.org/package/cryptonite-0.25/docs/Crypto-PubKey-ECC-ECDSA.html#v:sign) function also requires passing in a hash function. 
**[3]** `ETH` will be used to represent [Ethereum's address format](https://kobl.one/blog/create-full-ethereum-keypair-and-address/), or `BTC-P2PKH`, `BTC-P2SH` and `BTC-Bech32` to represent [Bitcoin's different address formats](https://en.bitcoin.it/wiki/Address).

**SUMMARY/TODOs:**
- [x] Change type of PPKScheme to include Address Format, Hash Algorithm, and Signature Algorithm. See above Motivation section for rationale.
- [x]  Decouple the different uses of `hash` in Pact. The same `hash` function handled hashing for Pact's native `hash` function, hashing transactions for acquiring the Request Keys for Commands, and hashing transactions to pass them to signing/validation Crypto functions.
- [x] Create own Pact signature data type. Pact has been using ED25519's Signature data type, but signature algorithms in Haskell output/expect Signatures in different formats (See [ED25519](http://hackage.haskell.org/package/ed25519-donna-0.1.1/docs/src/Crypto-Ed25519-Pure.html#Signature) vs [ECDSA's](http://hackage.haskell.org/package/cryptonite-0.25/docs/Crypto-PubKey-ECC-ECDSA.html#t:Signature) Signature data type). 
- [x] Create Pact's own Public/Private Key data type, as well as own import/export functions. Currently using ED25519's implementation, which presents similar issues as the above (See [ED25519](http://hackage.haskell.org/package/ed25519-donna-0.1.1/docs/src/Crypto-Ed25519-Pure.html#PublicKey)'s vs [ECDSA's](http://hackage.haskell.org/package/cryptonite-0.25/docs/Crypto-PubKey-ECC-ECDSA.html#t:PublicKey) Public Key implementation).
- [x] Add ECDSA to available signature schemes.
- [x] Where to inject formatted address to demonstrate ownership for ETH ledgers.
 
EDIT 12/07/2018:
The changes to `scheme` will be a breaking change for anyone that has been using `mkCommand` directly or who have written their own "make command" function.

EDIT 12/19/2018:
There are also breaking changes to `mkCommand`, PublicKey, PrivateKey, and ApiReq's `mkExec` and `mkCont`.

EDIT 01/15/2018:
The previous description assumed a sum type implementation. But after Dec 20th, the commits reflect attempts to use type classes instead.

slpopejoy commented on src-ghc/Pact/Types/Crypto.hs:
```diff
@@ -15,116 +17,268 @@
+  def = PPKScheme (def, def, def)
+
+
+data AddressFormat = Chainweb
```
AddressFormat "PubKey" for non-hashed PK

slpopejoy commented on src-ghc/Pact/Types/Crypto.hs:
```diff
@@ -15,116 +17,268 @@
   deriving (Show, Eq, Ord, Generic)
 
 instance NFData PPKScheme
 -- default instance with only one value is empty array!!
```
remove comment everywhere

slpopejoy commented on src-ghc/Pact/Types/Crypto.hs:
```diff
@@ -15,116 +17,268 @@
+         algorithms (i.e. ECDSA).
+      3. The signature algorithm used in signing transactions (i.e. ED25519, ECDSA)
+--}
+newtype PPKScheme = PPKScheme (AddressFormat, HashAlgo, SignatureAlgo)
```
Revert PPKScheme to being API enum and call this something else

slpopejoy replied:
PPKImplementation

slpopejoy commented on src-ghc/Pact/ApiReq.hs:
```diff
@@ -51,12 +51,25 @@ import Pact.Types.Runtime hiding (PublicKey)
   _kpSecret :: PrivateKey,
-  _kpPublic :: PublicKey
+  _kpPublic :: PublicKey,
+  _kpScheme :: Maybe AddressFormat
```
Go back to PPKScheme type here

slpopejoy replied:
Keep `ED25519` for current scheme

slpopejoy replied:
ED25519 = Blake512B + ED25519 + Full addys
CW = Blake256B + ED25519 + Blake128B addys?
ETH = ... + secp256k1 + eth addys

mightybyte commented on src-ghc/Pact/Types/Crypto.hs:
```diff
@@ -15,117 +17,168 @@
+
+
+
+class (Export (PublicKey a), Export (Signature a),
```
At a high level this looks like a pretty reasonable start to me.  I'd really like to see at least one more hash scheme implemented before passing judgment to see how things fall out.

LindaOrtega replied:
Like adding another PPKScheme? i.e. `PPKScheme = ED25519 | ETH`?

LindaOrtega commented:
(Adding this here for later)
Keyset enforcement is tied to how Crypto public keys are represented as Pact Runtime PublicKeys:
https://github.com/kadena-io/pact/blob/master/src-ghc/Pact/Types/Server.hs#L68

This would impact the UI/UX experience of people using YAML files or creating their own Exec commands since the value of "admin-keyset" will not correspond 1:1 with the value of the PublicKey being used to sign the transaction for ETH/BTC.

LindaOrtega commented on src-ghc/Pact/Types/Command.hs:
```diff
@@ -97,23 +97,22 @@ instance (FromJSON a) => FromJSON (Command a) where
 mkCommand :: (ToJSON m, ToJSON c) =>
              [SomeKeyPair] -> m ->
-             Text -> PactRPC c -> Command ByteString
+             Text -> PactRPC c -> IO (Command ByteString)
```
I needed to make this IO because signing in ECDSA adds some randomness, which ED25519 does not do.

slpopejoy commented on src-ghc/Pact/Types/ECC.hs:
```diff
@@ -0,0 +1,214 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Pact.Types.ECC
```
Should probably call this module ECDSA

slpopejoy commented on src-ghc/Pact/Types/ECC.hs:
```diff
@@ -0,0 +1,214 @@
+
+
+
+--------- ECDSA TESTS ---------
```
Should move these into unit tests if they are actually useful, or delete if they're not useful

LindaOrtega commented on src/Pact/Types/Command.hs:
```diff
@@ -133,39 +147,40 @@ verifyCommand orig@Command{..} = case (ppcmdPayload', ppcmdHash', mSigIssue) of
 
 #else
 
+-- TODO: Add ETH Scheme for GHCJS
```
This was my mistake when I was resolving the merge conflict. PPKScheme does work with either GHC or GHCJS. I've refactored PPKScheme into its own module (see 41b150c commit) so there's only one PPKScheme to update for both GHCJS and GHC.

LindaOrtega commented:
Regarding my previous [comment](https://github.com/kadena-io/pact/pull/338#issuecomment-455255635) and Stuart's review [comments](https://github.com/kadena-io/pact/pull/338#pullrequestreview-197329498): 
I've added a `format-address` function to the REPL (not GHCJS compatible) to make it easier for the user and devs to test that Crypto PublicKeys converted to Addresses (also known as Pact Runtime Public Keys) as expected. 
Moreover, I recently found that its trivial to deduce the Public Key (and hence the Address) from the Private Key for [Elleptic Curve Crypto schemes](https://stackoverflow.com/questions/696472/given-a-private-key-is-it-possible-to-derive-its-public-key). In fact, this seems to be what Ethereum's [GETH](https://kobl.one/blog/create-full-ethereum-keypair-and-address/#complete-example) is doing as it only requires a private key to create an account and returns to the user its address. 

Given this possibility, I changed `ApiReq` to only require a Private Key for schemes that allow for trivial derivation of Public Key. Public Key is an optional parameter. I also added an optional `Address` field to the keyset section of `ApiReq` . The module enforces that the provided address is the same as the address derived from the Public Key to ensure transparency. 

Moreover, this means that users only need to provide their address and their private key when signing transactions, which is a similar experience to the one they currently have with Ethereum.

marklnichols commented:
@Linda, are there a bunch of changes we should make to Kadena now re this public/private key stuff?

LindaOrtega commented:
> @linda, are there a bunch of changes we should make to Kadena now re this public/private key stuff?

If Kadena uses the [mkApiReq](https://github.com/kadena-io/pact/pull/338/files#diff-753a35b1e15d008e2a2e83df97e3778aR129) function then no changes will need to be made.

Changes will need to be made if any of the following functions are used:

1. [mkApiReqExec](https://github.com/kadena-io/pact/pull/338/files#diff-753a35b1e15d008e2a2e83df97e3778aR100) and [mkApiReqCont](https://github.com/kadena-io/pact/pull/338/files#diff-753a35b1e15d008e2a2e83df97e3778aR129) now take `[SomeKeyPair]` as an argument.

2. [mkExec](https://github.com/kadena-io/pact/pull/338/files#diff-753a35b1e15d008e2a2e83df97e3778aR118) and [mkCont](https://github.com/kadena-io/pact/pull/338/files#diff-753a35b1e15d008e2a2e83df97e3778aR154) now take `[SomeKeyPair]` instead of `[KeyPair]`.

3. [mkCommand](https://github.com/kadena-io/pact/pull/338/files#diff-07f1972f68441e4bcfba09fc7666f694R128) now takes `[SomeKeyPair]` instead of `[(PPKScheme, PrivateKey, Base.PublicKey)]`. It also now returns `IO (Command ByteString)` instead of `Command ByteString`.

4. Or any place that used PublicKey, PrivateKey, or KeyPair directly.

5. Or if someone created their own `mkCommand` function as UserSig has an extra [Address](https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src/Pact/Types/Command.hs#L247) field and the signatures for Crypto's `sign` and `verify` functions have changed.

I've added some "setter" and "getter" functions for working with `SomeKeyPair`. They can be generated and imported via:
https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L301

https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L310

https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L142


Each field of the `SomeKeyPair` can also be accessed via:
https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L288

https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L291

https://github.com/kadena-io/pact/blob/a6e924993a6770a9451b3f818be5feb634b977ea/src-ghc/Pact/Types/Crypto.hs#L294

slpopejoy commented on src-ghc/Pact/Types/Crypto.hs:
```diff
@@ -1,131 +1,337 @@
+  type PrivateKey a
+  type Signature a
+
+  _hashPayload :: a -> ByteString -> ByteString
```
Let's go back to using _cmdHash and not generating the hash as part of crypto

slpopejoy commented on src/Pact/Types/Command.hs:
```diff
@@ -133,39 +150,33 @@ verifyCommand orig@Command{..} = case (ppcmdPayload', ppcmdHash', mSigIssue) of
     parsePact code = ParsedCode code <$> parseExprs code
-    (ppcmdSigs' :: [(UserSig,Bool)]) = (\u -> (u,verifyUserSig _cmdHash u)) <$> _cmdSigs
-    ppcmdHash' = verifyHash _cmdHash _cmdPayload
+    (ppcmdSigs' :: [(UserSig,Bool)]) = (\u -> (u,verifyUserSig _cmdPayload u)) <$> _cmdSigs
```
Go back to using _cmdHash

mightybyte commented on src-ghc/Pact/ApiReq.hs:
```diff
@@ -41,22 +43,21 @@ import Data.Text.Encoding
+  _akpAddress :: Maybe Text,
+  _akpScheme  :: Maybe PPKScheme
+  } deriving (Eq, Show, Generic)
+instance ToJSON ApiKeyPair where toJSON = lensyToJSON 4
```
What context will these JSON serializations be used?  Will they be persisted in a way that requires backwards compatibility?  If so, then we might consider writing the JSON instances by hand to avoid upstream changing it on us without warning.

LindaOrtega replied:
They are used in `pact -a` and in the PactContinuationSpec. It essentially makes it easier to format API request JSON using a YAML file describing the request.

See: https://pact-language.readthedocs.io/en/latest/pact-reference.html#api-request-formatter

mightybyte replied:
Ok, this is probably not a big deal.

slpopejoy commented:
@joelburget @bts @mightybyte we're really getting killed on the CI failures, can we fix with some priority please?

@LindaOrtega can you do manual CI and report here that it passed? We'll merge after that.

LindaOrtega commented:
@slpopejoy I merged in the latest changes from master and ran `stack test` and `nix-build`. I got the same error that's shown on Gitlab:
```
tests/ClientSpec.hs:58:5:
  1) Client, Servant API client tests, correctly runs a simple command locally
       expected: Right (CommandSuccess {_csData = Number 3.0})
        but got: Left (ConnectionError "HttpExceptionRequest Request {\n  host                 = \"localhost\"\n  port                 = 8080\n  secure               = False\n  requestHeaders       = [(\"Accept\",\"application/json;charset=utf-8,application/json\"),(\"Content-Type\",\"application/json;charset=utf-8\")]\n  path                 = \"/api/v1/local\"\n  queryString          = \"\"\n  method               = \"POST\"\n  proxy                = Nothing\n  rawBody              = False\n  redirectCount        = 10\n  responseTimeout      = ResponseTimeoutDefault\n  requestVersion       = HTTP/1.1\n}\n (ConnectionFailure Network.Socket.connect: <socket: 22>: does not exist (Connection refused))")
```
